### ISSUE TITLE
Handling batch sizes

### ISSUE THREAD

**Issue by vmoens** [joecummings/r1-zero#11]
This task describes 2 items we need to achieve sequentially:

1. We must ensure that the trainer sees batches of data that originate from diferent prompts. In other words, the trainer batch (size) needs to be decoupled from the GRPO batch (size)
2. We need to take into account that the GRPO batch may come as separate items to the buffer, not necessarily in one go.

(2) is the end goal but harder to achieve than (1) so we need (1) to be done first, and (2) will follow.

## Naming

- A *trajectory** should be a single rollout that follows a prompt. If there are N repetitions for advantage computation, these N repetitons are N trajectories
- A *GRPO batch* is a number of repetitions of the same prompt
- A *training batch* is a bunch of trajectories for training

## Current situation

We currently sample a single prompt, run a GRPO batch in vLLM, send that GRPO batch to the post-proc, send that post-proc GRPO batch to the trainer. The trainer receives exactly the same number of elements as the GRPO batch.

## Eventually

- We need to assume that the collector gives us trajectories that are not batched. In other words, if the collector has ended the repetition `i` of `N` (N = GRPO batch size), it writes it in the buffer. That also means that 2 consecutive samples in the buffer may not originate from the same GRPO batch. This is because we want to be able to get large amounts of trajectories for a single prompt (we need to untangle the number of repetitions from the compute budget of inference).
- We don't want to overfit to the concept that the advantage is computed over a set of repetitions of one sample. That should be well isolated.

A first vestion of the flow could be
So here's the flow
```python

# collector loop
while True:
    prompt = next(dataloader)
    grpo_batch = []
    for i in range(grpo_batch_size):
        data = vllm_policy(prompt)
        grpo_batch.append(data)
    queue.add(grpo_batch)

# postproc / scorer loop
while True:
    grpo_batch = queue.popleft()
    # apply ref model, compute rewards, compute advantages
    # add every element of the GRPO batch independently - right now this is an "add", not an "extend", ie we consider that the batch is one element
    replay_buffer.extend(grpo_batch)

# trainer loop
while True:
    trainer_batch = replay_buffer.sample(trainer_batch_size)
    # execute loss, backprop etc

```

Eventually, we need to assume it won't be possible to pass the entire grpo batch in one go so we will need to be smart about this. 
IMO we will need for the replay buffer writer to keep track of the prompt of all the data it sees, and only once the number of repetitions of the same prompt is reached, the buffer does the following:

- The buffer computes the advantage
- The buffer assigns the advantage
- The buffer writes in its storage the fully processed data

**Comment by Darktex** [comment]
(When you say the buffer, I assume you mean the scorer :P )
